### ISSUE TITLE
Can't run two AppHost projects at the same time & warning from kestrel when running AppHost project

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/aspire#1639]
# Background

Now that the dashboard is split out from the AppHost and run in its own process, there are 3 endpoints in play by default when launching an AppHost project from VS, `dotnet run`, etc.:

- The URL to run the dashboard UI on. This is typically set via the `"applicationUrl"` property in the launch profile in the *launchSettings.json* file of the AppHost project, which is interpreted by the launcher (VS, `dotnet run`, etc.) and passed to the process via the `ASPNETCORE_URLS` environment variable. `Aspire.Hosting` then uses this as the URL to pass to the dashboard process (via `ASPNETCORE_URLS` env var) so that the dashboard UI can be browsed to via this URL. This is set by our AppHost templates by default to a random port in a specified range at project creation time, and falls back to a hard-coded value of `http://localhost:15888`.
- The URL to run the dashboard OTLP ingestion endpoint on. This is typically set via the `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` environment variable in the `"environmentVariables"` section in the launch profile in the *launchSettings.json* file of the AppHost project, which is interpreted by the launcher (VS, `dotnet run`, etc.) so that the defined environment variables are passed to the process. `Aspire.Hosting` then uses this as the URL to pass to the dashboard process so that the dashboard UI can be browsed to via this URL. This is set by our AppHost templates by default to a random port in a specified range at project creation time, and falls back to a hard-coded value of `http://localhost:15888`.
- **NEW**: The URL to run the AppHost process on. This serves as the gRPC resource server for the dashboard. This can bet set via the `DOTNET_DASHBOARD_GRPC_ENDPOINT_URL` environment variable which today isn't set by anything by default (e.g. our templates). The default fallback URL is hard-coded to `http://localhost:18999`.

As it stands right now, this results in not being able to run more than one AppHost project at a time by default, because the URL to run the AppHost resource server gRPC endpoint on is not set to anything, and thus always falls back to the hard-coded default of `http://localhost:18999`, resulting in TCP binding errors for everything but the already running AppHost process.

Additionally, because the `ASPNETCORE_URLS` environment variable is being used to pass the URL to use for the dashboard UI (see the first bullet above) to the AppHost project on launch, and Aspire.Hosting explicitly configures Kestrel for the AppHost process to listen on a different URL (last bullet above), Kestrel prints a warning saying that the URL passed via `ASPNETCORE_URLS` is being overridden.

# Proposed Changes

Rather than `Aspire.Hosting` falling back to hard-coded ports for the OTLP endpoint URL to configure the dashboard to host, and the resource server gRPC endpoint that the AppHost project hosts, it should instead fallback to randomly selected ports. These aren't user facing during the developer inner-loop as there's no need to interact with them directly, so using random ports by default and not having a static port configured in the templates seems like a reasonable default.

We should leave the capability in to enable setting these ports via environment variables on the AppHost process just in case there are scenarios where setting a specific port via the AppHost project might be useful (e.g. diagnosing issues, etc.).

Additionally, I propose that in addition to using a random port by default, we should match the scheme of the URL passed via `ASPNETCORE_URLS` (i.e. the AppHost project URL that's actually used as the dashboard URL). This way, if the AppHost is started with an `"applicationUrl"` launch profile property set to an HTTPS URL, the other 2 URLs will also be configured to use HTTPS with the random ports selected by `Aspire.Hosting`. Each of the three URLs can still be fully controlled individually if desired via the various environment variables described above.

RE the Kestrel warning from the AppHost process, we could choose to ignore it as it's only present during the developer inner-loop, but it **is** distracting and the number of URLs at play is confusing enough as it is. We could explicitly suppress it in the *appsettings.json* file of the AppHost project (in templates), but that would then mean users would need to manually enable it again to see warnings from Kestrel if they need to diagnose any issues with the resource server hosted by the AppHost. Interested in folks' thoughts on this piece.

**Comment by mitchdenny** [comment]
So concretely we'd have something like this:

```json
{
  "profiles": {
    "http": {
      "commandName": "Project",
      "launchBrowser": true,
      "dotnetRunMessages": true,
      "applicationUrl": "https://localhost:15888", // This port is randomly assigned at project creation, and hosts the resource server endpoint.
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development",
        "DOTNET_ENVIRONMENT": "Development",
        "DOTNET_DASHBOARD_UI_ENDPOINT_URL": "http://localhost:17361", //  New, overrides randomly generated default.
        "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL": "http://localhost:16031" // Overrides randomly generated default.
      }
    }
}
```

**Comment by davidfowl** [comment]
@drewnoakes can take a look?

**Comment by drewnoakes** [comment]
For sure. It's on my radar.

> Now that the dashboard is split out from the AppHost and run in its own process

To be clear, this is not yet the case. We have a gRPC endpoint to communicate data, but it currently runs in-proc. I'm currently working on splitting out the dashboard as a separate process.

**Comment by DamianEdwards** [comment]
@mitchdenny not quite. We possibly could do what you're suggesting, but then we'd have to update Visual Studio, `dotnet watch`, VS Code, etc. to understand that the `DOTNET_DASHBOARD_UI_ENDPOINT_URL` address is the one to launch the browser to (rather than `applicationUrl`), **OR** we also set the `launchUrl` property to the same value as `DOTNET_DASHBOARD_UI_ENDPOINT_URL`, which seems a bit duplicative.
